### PR TITLE
[WIP] Hierarchical Clustering

### PR DESCRIPTION
**Pull request by jtwinrelevanceai**
Added hierarchical clustering support for SDK, with dendrogram support for plotly/matplotlib

### CODE REVIEW

File: relevanceai/vector_tools/cluster.py [hunk 1]
```diff
@@ -332,6 +332,112 @@ def fit_transform(
...
+    ) -> np.ndarray:
+        try:
+            from sklearn.cluster import AgglomerativeClustering
+        except ModuleNotFoundError as e:
+            raise ModuleNotFoundError(
+                f"{e}\nInstall Hierarchical Clustering\n \
+                pip install -U relevanceai[hierarchical]"
+            )
+
+        agg = AgglomerativeClustering(
```
Comment (boba-and-beer) [inline relevanceai/vector_tools/cluster.py:368]: keep model as an attribute once fitted in case we need something else and don't want to re-fit

File: relevanceai/vector_tools/cluster.py [hunk 2]
```diff
@@ -332,6 +332,112 @@ def fit_transform(
...
+            )
+
+        vectors = np.array(vectors)
+        width = kwargs['width']
+        height = kwargs['height']
+        kwargs.pop('width')
+        kwargs.pop('height')
+        fig = ff.create_dendrogram(vectors, **kwargs)
+        fig.update_layout(width=width, height=height)
+        fig.show()
```
Comment (boba-and-beer) [inline relevanceai/vector_tools/cluster.py:423]: `return fig`

File: relevanceai/vector_tools/cluster.py [hunk 3]
```diff
@@ -332,6 +332,112 @@ def fit_transform(
...
+        except ModuleNotFoundError as e:
+            raise ModuleNotFoundError(
+                f"{e}\nInstall Hierarchical Clustering\n \
+                pip install -U relevanceai[hierarchical]"
+            )
+
+        vectors = np.array(vectors)
+        width = kwargs['width']
+        height = kwargs['height']
+        kwargs.pop('width')
```
Comment (boba-and-beer) [inline relevanceai/vector_tools/cluster.py:419]:
```
width = kwargs.pop('width')
height = kwargs.pop('height')
```

File: relevanceai/vector_tools/cluster.py [hunk 4]
```diff
@@ -627,3 +731,130 @@ def hdbscan_cluster(
...
+        # get vectors
+        if len(vector_fields) > 1:
+            raise ValueError(
+                "We currently do not support more than 1 vector field yet. This will be supported in the future."
+            )
+
+        if n_clusters is None and distance_threshold is None:
+            distance_threshold = 1
+        elif n_clusters is not None and distance_threshold is not None:
+            n_clusters = 10
```
Comment (boba-and-beer) [inline relevanceai/vector_tools/cluster.py:834]: If we set a default, then we should add a warning

File: setup.py [hunk 5]
```diff
@@ -48,6 +48,8 @@ def get_version(rel_path):
 umap = ["umap-learn>=0.5.2"]
 ivis_cpu = ["ivis[cpu]>=2.0.6"]
 ivis_gpu = ["ivis[gpu]>=2.0.6"]
+kmeans = ["scikit-learn>=1.0.1"]
+hierarchical = ["scikit-learn>=1.0.1"]
```
Comment (boba-and-beer) [inline setup.py:52]: nice

**Review by boba-and-beer — CHANGES_REQUESTED** [review]

File: relevanceai/vector_tools/cluster.py (hunk 4 above)
Comment (jtwinrelevanceai) [reply relevanceai/vector_tools/cluster.py:834]: `import warnings
warning.warning('setting distance_threshold=1`

like this?

File: relevanceai/vector_tools/cluster.py (hunk 3 above)
Comment (jtwinrelevanceai) [reply relevanceai/vector_tools/cluster.py:419]: true

File: relevanceai/vector_tools/cluster.py (hunk 4 above)
Comment (boba-and-beer) [reply relevanceai/vector_tools/cluster.py:871]: ^ `logger` class might better but not sure if that is default on?

**Comment by jtwinrelevanceai** [comment]
TODO:
 - add support for multiple vector searches in one endpoint
 - Insert centroids and then list closest to centers
 - OPTIONAL (fcluster)

**Comment by boba-and-beer** [comment]
Had a think through this. We actually support the first one via the second one (multiple vector searches are done to find the closest documents to the centroids. In your hierarchical clustering, you should insert the centroids for each cluster and then immediately run `list_closest_to_centroids`

**Comment by boba-and-beer** [comment]
TODO: 
- Switch to list closest to center 
- Make labels appear on the dendrogram 
- Use single dendrogram algorithms

**Comment by boba-and-beer** [comment]
We now have alternative hierarchical clustering methods.
